docs: Improve Kafka client tutorials (#1581)
redpanda-data/documentation: docs/develop/guide-go-cloud.mdx

---

> title: Build a Chat Room Application with Redpanda Cloud and Go

Author: ```suggestion
title: Build a Chat Room Application with Redpanda Cloud and Golang
```

> title: Build a Chat Room Application with Redpanda Cloud and Golang

Reviewer: Love how this continues the chat room app started in the QS--nice!

linkRoot: ../../
deployment: Cloud
contextLinks:
  - name: 'Docker'
    to: 'develop/guide-go'
  - name: 'Cloud'
    to: 'develop/guide-go-cloud'
---

<head>

>     <meta name="title" content="Build a Chat Room Application with Redpanda Cloud and Go | Redpanda Docs"/>

Author: ```suggestion
    <meta name="title" content="Build a Chat Room Application with Redpanda Cloud and Golang | Redpanda Docs"/>
```

>     <meta name="description" content="Create a basic chat room application with Redpanda Cloud and Kafka clients developed with kafkajs"/>

Author: ```suggestion
    <meta name="description" content="Create a basic chat room application with Redpanda Cloud and Kafka clients developed with franz-go"/>
```

<SetUpCloud frontmatter={frontMatter}/>

## Set up your environment

1. Create and set your working directory to the project folder, `chat-room`:

   ```bash
   mkdir chat-room
   cd chat-room
   ```

2. Initialize the project:

  ```bash
  go mod init com/redpanda/chat-room
  ```

3. Install the required dependencies:

  ```bash
  go get github.com/twmb/franz-go
  go get github.com/twmb/franz-go/pkg/kadm
  go get github.com/google/uuid
  go get github.com/twmb/tlscfg
  go get github.com/twmb/franz-go/pkg/sasl/scram@v1.9.0
  ```

## Create a topic

You need a topic named `chat-room` for both Redpanda and the client to use to store chat room messages. If you completed the Redpanda Quickstart, this topic already exists in your cluster.

1. Verify that the `chat-room` topic exists in your cluster by listing all topics:

  ```bash
  rpk topic list --tls-enabled
  ```

  Output:

  ```
  NAME       PARTITIONS  REPLICAS
  chat-room  1           1
  ```

1. If the topic doesn't exist yet, use [rpk](../../reference/rpk) to create a `chat-room` topic:

  ```bash
  rpk topic create chat-room --tls-enabled
  ```

  Output:

  ```
  TOPIC      STATUS
  chat-room  OK
  ```

> ## Make sure the topic exists on the client side

Reviewer: ```suggestion
## Confirm that the topic exists on the client side
```

}

func (c *Consumer) PrintMessages() {
	ctx := context.Background()

	for {
		fetches := c.client.PollFetches(ctx)
		iter := fetches.RecordIter()
		for !iter.Done() {
			record := iter.Next()

			var msg Message
			if err := json.Unmarshal(record.Value, &msg); err != nil {
				fmt.Printf("Error decoding message: %v\n", err)
				continue
			}
			fmt.Printf("%s: %s\n", msg.User, msg.Message)
		}
	}
}

func (c *Consumer) Close() {
	c.client.Close()
}
```

You now have a consumer that reads all messages from the `chat-room`
topic and prints them to the console. You can start as many consumer groups as
you like, but remember that each group reads a message only once, which is
why the example is using a generated UUID for the group ID. This way, each time you run the application, you see all previous messages.

## Create a client application

The client needs an application that creates the topic, producer, and consumer and implements the chat logic.

To create a client application, in `chat-room/` create a source file `main.go` with the following content.

Replace the placeholders with the same values that you used in the Redpanda Cloud Quickstart.

> ```go title="main.go"

Reviewer: NOTE
These sample files are based on the assumption that you're running the Redpanda Quickstart, where the external port for broker redpanda-0 is set to port 19092. If you're not using the Redpanda Quickstart, make sure that the broker's port is correct for your deployment.

</head>

import Intro from "../develop/shared/chat-app/_intro.mdx"

> import SetUpCloud from "../develop/shared/chat-app/_set-up-docker.mdx"

Reviewer 2: Wrong path here
```suggestion
import SetUpCloud from "../develop/shared/chat-app/_set-up-cloud.mdx"
```
